### ISSUE TITLE
Strange type inference with `TypeVarTuple` (PEP 646)

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#2662]
Version: from `main`, commit `275956d3d1e1814a3562a12a9c2a2ef4f3fde326`

## Code
```py
from typing import Callable, Generic, TypeVar, Union
from typing_extensions import TypeVarTuple, Unpack


P = TypeVarTuple("P")
T = TypeVar("T", covariant=True)


class Call(Generic[Unpack[P]]):
    def __init__(self, *args: Unpack[P]) -> None:
        self.args = args


class Return(Generic[T]):
    def __init__(self, /, result: T) -> None:
        self.result = result


TailRec = Union[Call[Unpack[P]], Return[T]]


def tail_rec(fn: Callable[[Unpack[P]], TailRec[Unpack[P], T]]) -> Callable[[Unpack[P]], T]:
    ...


@tail_rec
def factorial(n: int, acc: int) -> TailRec[int, int, int]:
    if n <= 0:
        return Return(acc)
    return Call(n - 1, acc * n)


reveal_type(factorial)
```

Expected: `Type of "factorial" is "(int, int) -> int"`
Got: `Type of "factorial" is "(int, int) -> T@tail_rec"`

**Comment by decorator-factory** [comment]
If you remove the return type annotation from `tail_rec` and replace it with this: ```py
def tail_rec(fn: Callable[[Unpack[P]], TailRec[Unpack[P], T]]):
    def new_fn(*args: Unpack[P]) -> T:
        ...
    return new_fn
```
 the final type for `factorial` is inferred correctly.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.194, which I just published. It will also be included in the next release of pylance.
